projects/plugins/boost/wp-js-data-sync.php

```diff
@@ -1,22 +1,26 @@
 <?php
 
-use Automattic\Jetpack\WP_JS_Data_Sync\Boolean_Entry;
 use Automattic\Jetpack\WP_JS_Data_Sync\Data_Sync;
 use Automattic\Jetpack\WP_JS_Data_Sync\Data_Sync_Entry;
+use Automattic\Jetpack\WP_JS_Data_Sync\Data_Sync_Option;
 use Automattic\Jetpack\WP_JS_Data_Sync\Registry;
-use Automattic\Jetpack_Boost\Data_Sync\Cloud_CSS_Sync;
-use Automattic\Jetpack_Boost\Data_Sync\Critical_CSS_State_Handler;
+use Automattic\Jetpack\WP_JS_Data_Sync\Schema\Schema;
 use Automattic\Jetpack_Boost\Lib\Status;
 
 if ( ! defined( 'JETPACK_BOOST_DATASYNC_NAMESPACE' ) ) {
 	define( 'JETPACK_BOOST_DATASYNC_NAMESPACE', 'jetpack_boost_ds' );
 }
 /**
  * Functions to make it easier to interface with Async Option:
+ *
+ * @throws Exception
  */
-function jetpack_boost_register_option( $name, $handler ) {
+function jetpack_boost_register_option( $key, $schema, $entry = null ) {
+	if ( ! $entry ) {
+		$entry = new Data_Sync_Option( JETPACK_BOOST_DATASYNC_NAMESPACE, $key );
+	}
 	return Registry::get_instance( JETPACK_BOOST_DATASYNC_NAMESPACE )
-					->register( $name, $handler );
+					->register( $key, new Data_Sync_Entry( $entry, $schema ) );
```

Comment:
Since the reason behind `Data_Sync_Entry` is to enforce the schema validation, and per my understanding it is not supposed to be customized, shouldn't this be an internal class and should the registry have an interface like:

```
public function register( $key, $schema, $entry )
```

And then, the registry instantiates the `Data_Sync_Entry` class as a wrapper of the $entry(an instance of `Data_Sync_Option` by default).

Also, it looks like the namespace needs to be injected in two places. Can we require the storage handlers to have a method like `setNamespace`, which the registry then uses to set the namespace while registering?

Reply:
> Also, it looks like the namespace needs to be injected in two places. Can we require the storage handlers to have a method like setNamespace, which the registry then uses to set the namespace while registering?

I think it's good to namespace options, which is why Data_Sync_Option requires a namespace upon instantiation.  However, not every Entry type requires a namespace, like in the example given in the PR description.

The code here might look like one namespace depends on the other, but that's not the case. I could, and probably should, and probably will, rename the option namespace to option prefix - it can be the same thing as the actual endpoint namespace, but the option namespace isn't related to it in any other way than just by name.

---
Side note: 

The reason for Data_Sync_Entry is not only to enforce a schema, but to also provide a common and predictable entry interface. 
---

As for moving more of the logic inside the register method - I think that's not the right approach - it would make "the registry" do more than just "register a reference" - it would also be responsible for setting up.

But your prodding made me realize here that Data_Sync_Entry as the argument is also incorrect because it requires a concrete class and not a contract. I'll make a few changes there to ensure that part is flexible and clear too.